rust/connlib/tunnel/src/client.rs

```diff
@@ -893,103 +893,97 @@ impl ClientState {
     /// This is important because we don't want to lose state like resolved DNS names for resources that didn't change.
     ///
     /// TODO: Add a test that asserts the above.
-    ///       That is tricky because we need to assert on state deleted by [`ClientState::remove_resources`] and check that it did in fact not get deleted.
+    ///       That is tricky because we need to assert on state deleted by [`ClientState::remove_resource`] and check that it did in fact not get deleted.
     fn set_resources(&mut self, new_resources: Vec<ResourceDescription>) {
-        self.remove_resources(
-            &HashSet::from_iter(self.resource_ids.keys().copied())
-                .difference(&HashSet::<ResourceId>::from_iter(
-                    new_resources.iter().map(|r| r.id()),
-                ))
-                .copied()
-                .collect_vec(),
-        );
+        for id in HashSet::from_iter(self.resource_ids.keys().copied())
+            .difference(&HashSet::<ResourceId>::from_iter(
+                new_resources.iter().map(|r| r.id()),
+            ))
+            .copied()
+        {
+            self.remove_resource(id);
+        }
 
-        self.add_resources(
-            &HashSet::from_iter(new_resources.iter().cloned())
-                .difference(&HashSet::<ResourceDescription>::from_iter(
-                    self.resource_ids.values().cloned(),
-                ))
-                .cloned()
-                .collect_vec(),
-        );
+        for resource in HashSet::from_iter(new_resources.iter().cloned())
+            .difference(&HashSet::<ResourceDescription>::from_iter(
+                self.resource_ids.values().cloned(),
+            ))
+            .cloned()
+        {
+            self.add_resource(resource)
+        }
     }
```

Comment:
Here I'm seeing this (as pseudocode):
- Clone `new_resources` even though we own it
- Put that into a `HashSet`
- Clone all our current resources and make those a `HashSet` too
- Diff them
- Add the resources that we don't already have

But since a `ResourceDescription` already contains its own ID, could we not do this and save a few string clones?
- Clone only the IDs of our current resources into a `HashSet<ResourceId>`
- Do `new_resources.into_iter().filter()` with the `HashSet<ResourceId>` to get an owned iterator of resources that we don't already have
- Add the resources that we don't already have

Reply:
I've tried to clarify the implementation with some refactoring, let me know what you think.

Reply:
I realised that we can simplify things here a bit.

Reply:
The new one is much easier to read! But I was not sure if `add_resource` is idempotent or not. But the ID does not tell us everything, it's actually an upsert, right? 

So:
- If something other than the ID changed, we remove it and add it back
- If nothing changed, we don't remove it, but we do call some lower-level `add` function which probably is idempotent?

Reply:
> * some lower-level `add` function which probably is idempotent?

Yes, we just update entries in a HashMap there. I think we can actually test this by calling `set_resources` as part of roaming in `tunnel_test`, let me fix that real quick.